src/groupeddataframe/splitapplycombine.jl

```diff
@@ -597,33 +606,55 @@ function combine_helper(f, gd::GroupedDataFrame,
+        hcat!(newparent,
+              select(valscat, Not(intersect(keys, _names(valscat))), copycols=false),
               copycols=false)
         ungroup && return newparent
 
-        if length(idx) == 0
+        if length(idx) == 0 && !(keeprows && length(keys) > 0)
             @assert nrow(newparent) == 0
             return GroupedDataFrame(newparent, collect(1:length(gd.cols)), Int[],
-                                    Int[], Int[], Int[], 0, Dict{Any,Int}())
+                                    Int[], Int[], Int[], 0, Dict{Any,Int}(),
+                                    Threads.ReentrantLock())
         end
         if keeprows
+            @assert length(keys) > 0 || idx == gd.idx
+            @assert names(newparent, 1:length(gd.cols)) == names(parent(gd), gd.cols)
             # in this case we are sure that the result GroupedDataFrame has the
-            # same structure as the source
-            # we do not copy data as it should be safe - we never mutate fields of gd
-            return GroupedDataFrame(newparent, gd.cols, gd.groups, gd.idx,
-                                    gd.starts, gd.ends, gd.ngroups, getfield(gd, :keymap))
+            # same structure as the source except that grouping columns are at the start
+            return GroupedDataFrame(newparent, collect(1:length(gd.cols)), copy(gd.groups),
+                                    maybe_copy(getfield(gd, :idx)),
+                                    maybe_copy(getfield(gd, :starts)),
+                                    maybe_copy(getfield(gd, :ends)),
+                                    gd.ngroups,
+                                    maybe_copy(getfield(gd, :keymap)),
+                                    Threads.ReentrantLock())
         else
             groups = gen_groups(idx)
             @assert groups[end] <= length(gd)
+            @assert names(newparent, 1:length(gd.cols)) == names(parent(gd), gd.cols)
             return GroupedDataFrame(newparent, collect(1:length(gd.cols)), groups,
-                                    nothing, nothing, nothing, groups[end], nothing)
+                                    nothing, nothing, nothing, groups[end], nothing,
+                                    Threads.ReentrantLock())
         end
     else
-        if ungroup
-            return keepkeys ? parent(gd)[1:0, gd.cols] : DataFrame()
+        if keeprows
+            if nrow(parent(gd)) > 0
+                 throw(ArgumentError("select and transform do not support " *
+                                    "`GroupedDataFrame`s from which some groups have "*
+                                    "been dropped (including skipmissing=true)"))
+            end
+            if ungroup
+                return keepkeys ? select(parent(gd), gd.cols, copycols=copycols) : DataFrame()
+            else
+                return groupby(select(parent(gd), gd.cols, copycols=copycols), 1:length(gd.cols))
```

Comment:
Why not build the `GroupedDataFrame` directly? Same below.

Reply:
OK - I will change it.